node/src/components/chainspec_loader.rs

```diff
@@ -538,83 +539,103 @@ fn next_upgrade(dir: PathBuf, current_version: ProtocolVersion) -> Option<NextUp
         assert!(!ChainspecLoader::should_exit_for_upgrade(
-            highest_block,
+            highest_block_header,
             next_upgrade_activation_point
-        ),);
-
-        let highest_block = Some(Box::new(Block::random_with_specifics(
-            &mut rng,
-            EraId::from(2),
-            HEIGHT,
-            PROTOCOL_VERSION,
-            IS_SWITCH,
-            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
-        )));
+        ));
+
+        let highest_block_header = Some(Box::new(
+            Block::random_with_specifics(
+                &mut rng,
+                EraId::from(2),
+                HEIGHT,
+                PROTOCOL_VERSION,
+                IS_NOT_SWITCH,
+                VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+            )
+            .header()
+            .clone(),
+        ));
         let next_upgrade_activation_point = None;
         assert!(!ChainspecLoader::should_exit_for_upgrade(
-            highest_block,
+            highest_block_header,
             next_upgrade_activation_point
         ));
 
-        let highest_block = None;
+        let highest_block_header = None;
         let next_upgrade_activation_point = Some(ActivationPoint::EraId(10.into()));
         assert!(!ChainspecLoader::should_exit_for_upgrade(
-            highest_block,
+            highest_block_header,
             next_upgrade_activation_point
         ));
 
-        let highest_block = Some(Box::new(Block::random_with_specifics(
-            &mut rng,
-            EraId::from(2),
-            HEIGHT,
-            PROTOCOL_VERSION,
-            IS_SWITCH,
-            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
-        )));
+        let highest_block_header = Some(Box::new(
+            Block::random_with_specifics(
+                &mut rng,
+                EraId::from(2),
+                HEIGHT,
+                PROTOCOL_VERSION,
+                IS_NOT_SWITCH,
+                VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+            )
+            .header()
+            .clone(),
+        ));
         let next_upgrade_activation_point = Some(ActivationPoint::EraId(3.into()));
         assert!(!ChainspecLoader::should_exit_for_upgrade(
-            highest_block,
+            highest_block_header,
             next_upgrade_activation_point
         ));
 
-        let highest_block = Some(Box::new(Block::random_with_specifics(
-            &mut rng,
-            EraId::from(2),
-            HEIGHT,
-            PROTOCOL_VERSION,
-            IS_SWITCH,
-            VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
-        )));
+        let highest_block_header = Some(Box::new(
+            Block::random_with_specifics(
+                &mut rng,
+                EraId::from(2),
+                HEIGHT,
+                PROTOCOL_VERSION,
+                IS_NOT_SWITCH,
+                VERIFIABLE_CHUNKED_HASH_ACTIVATION.into(),
+            )
+            .header()
+            .clone(),
```

Comment:
There's also a `Block::take_header` function.

Reply:
Good tip - I'll remember for the future, but leave this as is, since it's only in a test.